### PR TITLE
Monkey recycler connected to xenobio consoles, recycler returns reduced

### PR DESCRIPTION
**Pull request by kriskog**
[Changelogs]: # (Your PR should contain a detailed changelog of notable changes, titled and categorized appropriately. This includes, new features, sprites, sounds, balance changes, admin tools, map edits, removals, big refactors, config changes, hosting changes and important fixes. An example changelog has been provided below for you to edit. If you need additional help, read https://github.com/tgstation/tgstation/wiki/Changelogs)

:cl: Skoglol
fix: Monkey recyclers now connected to the xenobiology consoles. Upgrading it now increases yield from monkeys gathered through the camera.
balance: Adjusted monkey recycler numbers. Down from 4 cubes to 0.8 cubes per monkey at t4 parts. 
/:cl:

[why]: # (Please add a short description [two lines down] of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.)

Monkey recyclers aren't currently used by the consoles. Upgrading parts does not increase the amount of monkey you get back from recycling.

Rewrote the math for parts, the old way of calculating it was confusing. Reduced amount of monkeys returned greatly, it makes little sense to quadruple your input monkey mass after the slimes are done eating. Considering most players have been making do with 0.2 back per monkey, 0.8 should be plenty. Open for feedback on this.

### CODE REVIEW

**Comment by Cyprex** [comment]
Early on where you're mostly waiting for slimes to feed+split and need the cubes the recycler is very useful, much less so later on due to Monkey cubes being abundant(Reproductive grey, industral grey or simply just grab a monkey for easy blood). As such, the recycler needs to be very good later on. I am against linking it to the console, as I believe it is fine as is and having the choice between an easy and a more labor-intensive but also more rewarding method is good. If anything, up the monkey gain for lower tiers a bit.

**Comment by kriskog** [comment]
Well it is a recycler, not a magic monkey replication machine. Or rather, currently it is the latter but the name implies it should be the former.
If you want to increase the amount of monkey cubes you have, you should be using the reproductive greys or breeding more greys as the labor intensive but more rewarding choice. 

As for early game returns, you get t3 parts 10 minutes into most rounds, which would yield 0.6 monkeys per monkey recycled. At that point, you should have had time to split slimes once (6 monkeys used), which leaves you four monkeys to start growing the next batch of slimes. And that is just the starting cubes, you can easily make new ones.
I am not wholly against increasing the starting amount, but I am not sure it is strictly needed.

File: code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm
Comment (nicbn) [inline code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm]: why

Comment (kriskog) [reply code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm]: It no longer applies

**Comment by kriskog** [comment]
> 
> 
> Yeah, the ratio is fine as is. If this PR got merged there would be no reason whatsoever to make industrial grey extracts, as the need for monkey cubes simply is not there, considering how you get back free monkeys easily


I am honestly not entirely sure what you want. You seem to be contradicting your own statements. 

I do not agree that leaving the recycler unlinked from the consoles is the right way to go forward. Leaving the ratio where it was (1 monkey per 4 cubes) is also clearly not an option, as you state that even the new ratio (1 monkey per 0.8 cubes) is too much. So clearly, the ratio is not fine as is.

File: code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm
Comment (ExcessiveUseOfCobblestone) [reply code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm]: lol this post

File: code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm]: ?

**Comment by kriskog** [comment]
>
> 
> He is, but a point matching mine can be gleaned from it: right now you can either spend your time dragging monkeys, or you can use an extract instead to make them; with your (misguided) change to recycler efficiency scientists are _forced_ to use an extract, removing one of the options, because the other no longer has any benefit.

At t3 parts, which you can easily get 10 minutes in, you'll be getting back 3 out of 5 monkeys. It's not like you're suddenly going to spend all your time making monkeys, but yes you will have to make some. As you should. I for one never used the recycler manually, and easily got by with the old console ratio of 0.2 monkeys back per monkey. 
Arguing that the work of manually grinding a monkey should be worth at most 20 times that is silly.

> > Rewrote the math for parts, the old way of calculating it was confusing.
> 
> It's not confusing, it's integers, because it lets players not do basic math.

The old function yielded an exponential increase, the new one yields a simple linear increase. 



Going forward, please supply actual reasonable arguments for why the rate is off if you wish to discuss it. "I dont like change" and "You are fixing a thing that I liked when broken" are not constructive arguments.

**Comment by YakumoChen** [comment]
Why does turning a monkey into a cube produce less or more monkey in the first place? What defines the amount of monkey a monkey contains needed to make a cube?

Why is having the recycler produce extra cubes per monkey bad?

**Comment by Barhandar** [comment]
>As you should. 

No. 
>I for one never used the recycler manually, and easily got by with the old console ratio of 0.2 monkeys back per monkey.

Then you shouldn't be touching the monkey balance because you're _blatantly_ biased.
>The old function yielded an exponential increase, the new one yields a simple linear increase.

And how is linear increase good?
>are not constructive arguments.

Neither is "I managed to work with this fine". You're removing a valid option _because you don't like it_.

**Comment by kriskog** [comment]
> > are not constructive arguments.
> 
> Neither is "I managed to work with this fine". You're removing a valid option _because **you** don't like it_.

I am reducing the output of recycling because it mechanically makes no sense. You are turning a net positive for recycling at t3, at that rate we may as well just move away from having a finite amount of monkeys in the first place.

**Comment by DfourC** [comment]
Shouldn't the name of the PR mention that it drastically changes the output of the recycler?

File: code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm
Comment (kriskog) [reply code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm]: The list holds references to all the consoles that is connected to the recycler. It is used when the recycler is destroyed for some reason, removing the reference from the consoles.

**Comment by drderp3635** [comment]
>recyler bad b cuz net positive gain 
>everything that makes monkeys produces a net positive in a closed system 

Pr is bad and you should feel bad

File: code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm
Comment (ExcessiveUseOfCobblestone) [reply code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm]: then it should be `var/list/connected`

**Comment by ExcessiveUseOfCobblestone** [comment]
Monkeys literally are infinite with T4 tho lol

**Comment by davethwave** [comment]
You do realize that you only get a max of .8 per monkey with this system? Its not  infinite or anything like that.

**Comment by ExcessiveUseOfCobblestone** [comment]
you get 400% return with T4 as of current, sorry for not being clear.

File: code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm
Comment (kriskog) [reply code/modules/food_and_drinks/kitchen_machinery/monkeyrecycler.dm]: Fair enough, I will change it.

**Comment by epochayur** [comment]
xeno mains really bringing the butthurt

**Comment by Poojawa** [comment]
why is sitting in one spot playing an arcade grabber machine considered fun?

Why does it need more automation?

**Comment by kriskog** [comment]
> 
> 
> why is sitting in one spot playing an arcade grabber machine considered fun?
> 
> Why does it need more automation?

Recycling monkeys via the console is already a thing most people use.

> 
> 
> Ah, the good ol' "lets bundle a massive nerf together with a QoL change to make it more likely to get merged" strategy.
> 
> -1

This QoL change cannot be made without getting rid of the old loophole of using the recycler manually. Monkeys being infinite 6 minutes into the round does not seem intended to me. Soon as you got t3, you can start killing monkeys to multiply them. Xenobiology has lots of other ways of making monkeys that are made irrelevant by way too high returns from the _**recycler**_.


I would like some maintainer opinions on this nerf. If I am way off base here, let me know and I will revert it.

**Comment by Barhandar** [comment]
>Recycling monkeys via the console is already a thing most people use.

Bullshit. Unless you can provide statistics, you're pulling this out of your arse.
>This QoL change cannot be made without getting rid of the old loophole of using the recycler manually. 

It is not a loophole. And it can be made just fine, beyond fine even because _you cannot pop the cubes out of the console_, which means either the monkeys are recycled into the console, or manually in the recycler if you want to do _anything_ like reproductive extracts.
>Xenobiology has lots of other ways of making monkeys 

Yes, and removing one of them is bad.
>that are made irrelevant by way too high returns

They're usable without monkeys and without recycler, unlike this method. And see above about your own linking invalidating your point.
>Monkeys being infinite 6 minutes into the round

They're infinite 30 seconds into the round. Also, that only applies if scientists _beeline_ to H/E parts, while usually mining tech comes first.

**Comment by ExcessiveUseOfCobblestone** [comment]
I will say I use the console because the slime pro gives you more than enough grey slimes to work with. I just buckle a monkey down to a bed I make with the plentiful metal around science and siphon blood for the reaction, it certainly appears to be a lot faster than manually going into pens to pull out monkeys, even if you relocate everything which is a timesink in itself.

File: code/modules/research/xenobiology/xenobio_camera.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/research/xenobiology/xenobio_camera.dm]: Where's the crazy math that requires this safeguard or perhaps what's the reason behind it?

If you're only working with tenths it seems pretty silly to "make sure" you're working with tenths.

File: code/modules/research/xenobiology/xenobio_camera.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/research/xenobiology/xenobio_camera.dm]: Similar to above, I'm not seeing the crazy math that requires this safeguard.

File: code/modules/research/xenobiology/xenobio_camera.dm
Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: Floats is the reason. It would cause chat outputs like: "[console] now has 0.60000032  monkeys available. "
I suppose it should strictly speaking be using rounding instead of flooring, since it can go both ways. Other option is formatting the chat output, but I rather like neat numbers in my vars.

I'll replace the floors with round instead though.

File: code/modules/research/xenobiology/xenobio_camera.dm
Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: Same as above

File: code/modules/research/xenobiology/xenobio_camera.dm
Comment (ExcessiveUseOfCobblestone) [reply code/modules/research/xenobiology/xenobio_camera.dm]: I'm asking what math are you doing to get the 0.6**000000032**

Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: Unlike integer types, floats are inaccurate and often cause issues like this. This is not DM specific, it is just a shortcoming of floating point numbers in code.

This site explains it well if you are interested: [link](http://effbot.org/pyfaq/why-are-floating-point-calculations-so-inaccurate.htm)

Comment (Barhandar) [reply code/modules/research/xenobiology/xenobio_camera.dm]: Floats are inaccurate at high values. Have you actually TRIED and saw that it doesn't truncate, or are you just applying "good coding practices" at random?

Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: @Barhandar This was a fix for that specific issue, as you can see in the commit history. Floats are inaccurate at all values.

@ExcessiveUseOfCobblestone You can't represent 0.8 in integers. Does DM even have integers? I thought I read it was floats only, but I might be mixing it up with something else.

**Comment by optimumtact** [comment]
Frankly I'm not very pleased with infinite monkey's as a concept at all.

**Comment by kriskog** [comment]
> Frankly I'm not very pleased with infinite monkey's as a concept at all.

Right, good? You're a bit unclear, do you think 0.8 monkeys returned is too high for T4?

**Comment by ExcessiveUseOfCobblestone** [comment]
0.8 should be fine.

**Comment by davethwave** [comment]
Indeed. Frankly 4 monkey cubes for a monkey corpse was always ridiculous in my opinion.

**Comment by kriskog** [comment]
Any updates on this getting merged? I am working on something else xenobio that will cause merge issues with this.

**Comment by Cyprex** [comment]
There are no guarantees this will be merged at all, and I hope it won't.

**Review by ExcessiveUseOfCobblestone — DISMISSED** [review]

File: code/modules/research/xenobiology/xenobio_camera.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/research/xenobiology/xenobio_camera.dm]: Will this ever not be an integer?

Previously eating monkeys gave the console itself +0.2 monkeys, but now you HAVE to use the recycler for cubes which will always give +1 monkey when inserted into a console

Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: You seem to be under the misconception that we add the monkeys to the recycler though the console and grind them, and that isn't really what's happening. This is just using the parts from the recycler to upgrade the recycling through the console as well as the manual grinding. The increase in monkeys available happens just like before (though higher numbers with upgrades), and will be a decimal number most of the time. 

I believe we discussed that point in a previous review, although it was when we picked them up. Due to giving chat output directly from this variable, it needs to round or you'll get imprecise numbers output to chat. Now we might be okay only rounding on pickup, but if I am being honest, it's been almost a month so I no longer remember precisely what I tested to fix the rounding errors.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/research/xenobiology/xenobio_camera.dm]: Currently:
- Suck up monkey via console
- Console gets 0.2
- Recycler gets 0

- Dispense Monkey via console
- Console loses 1
- Recycler loses 0 (not including cube dispensed from recycler)

PR:
- Suck up monkey via console
- Console gets 0
- Recycler gets (math in tenths)

- Dispense Monkey via console
- Console loses 1
- Recycler loses 0 (not including cube dispensed from recycler)

The ONLY way to get monkeys via console now is to use cubes which ONLY give pure whole numbers.

Basically i'm suggesting turning the second arg into 1 or dropping it. I have no issue with the round thing since you explained that already. Does that make sense lol?

Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: You still got it wrong. There are two things happening in this PR, not one. The monkey recycler has had the parts changed to do different stuff than before, to give the new ratio of 0.2-0.8 and 
 to be equal to using the consoles recycle action.
The console uses the production var from the recycler to check how many monkeys are gained when pulling up a monkey, and adds this to the consoles monkeys var.

At no point does the console add stored_matter to the recycler, or vice versa.

So sucking up a monkey after the PR:
- Suck up monkey via console
- Console gets 0.2-0.8 monkeys
- Recycler gets 0

- Dispense monkey via console
- Console loses 1

In short, the monkeys are added directly to the console and never really visits the recycler unless you do it manually.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/research/xenobiology/xenobio_camera.dm]: correct, this is why you don't second guess your first review 8)

File: code/modules/research/xenobiology/xenobio_camera.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/research/xenobiology/xenobio_camera.dm]: This makes it so ANY xenobio cam syncs to this recycler (including say ones in lavaland or space).

Consider a check similar to the recycler here.

Sorry for being nitpicky I wanna merge this but don't want to get bullied if something is wrong l0l

Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: That's a fair point, I'll make the glob a list, and iterate through and check that they are in the same area as the console before adding them. This will still allow manual connections, but roundstart initialization should never connect the wrong thing.

Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: Actually, looking back I made it so that only a monkey recycler on the station z-level is added to the glob. I'll still make the change though.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/research/xenobiology/xenobio_camera.dm]: correct, i'm just referring to the cams.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/research/xenobiology/xenobio_camera.dm]: get_area returns the area instance so you don't have to worry about name.

Area names might be duplicated, I haven't checked blueprints/make a new area to see if they allow for that.

Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: Right, yeah that makes sense. I am still thinking a bit backwards about the DM code in some cases, but this is indeed comparing the area instance which would make it unique.

This should be good to go then, I tested and it works for on-station consoles/recycler. There are currently not any xenobio setups that initialize on mapload anywhere else, so this is harder to test. Still, if it works on station it should work everywhere else too.

**Comment by Cyprex** [comment]
This PR will remove the interesting choice between the labor-intensive, but more rewarding choice of manually dragging monkeys into the recycler and using the console. Additionally, with a fully upgraded recycler(~25m into the round) you only need 0.2 cubes per monkey, as you get 0.8 back from recycling. This effectively means you will need to produce a lot less monkey cubes.
If #41575 is merged xenobio also simply becomes too easy, as this PR is currently a huge buff to people who don't use the recycler manually.
tldr; There are 2 playstyles, use recycler and don't use the recycler, this removes the first and massively buffs the second.

**Comment by Cyprex** [comment]
To remedy the issues outlined above, I propose the following: 
- Recycler returns 0.5/1.0/1.5/2.0 monkeys per dead monkey inserted.
- Using the consoles built-in recycling returns 20% of what you have gotten in the linked recycler

These changes would make manually using the recycler still worth it and would put us at 0.1/0.2/0.3/0.4 cubes per monkey, which is still a considerable buff in my opinion. The numbers are obviously up to debate, and I'd like to hear your thoughts on this.

**Comment by kriskog** [comment]
Since #41575 got merged, I need to update this one slightly to include the new stuff from there.

**Comment by optimumtact** [comment]
I think I will merge this, since it's now easier to do xeno biology with the shortcuts

**Comment by kriskog** [comment]
This is now updated to also cover the changes in #41575.

**Comment by Cyprex** [comment]
@optimumtact Are you aware that this PR, as it is, will make xenobio a lot easier? Again, with the current numbers, you get 0.6 monkeys back per monkey with just T3 parts. Put differently, each feeding only costs 0.4 Monkey cubes, as you get 0.6 back. This gets even more ridiculous with T4 parts.
@kriskrog
I've tried semi-automatic monkey recycling with current T4 parts in various setups(conveyor+grille, tile with high pressure plasma surrounded my holofirelocks etc.) and come to the conclusion that it's not a big issue.
Anyway, I still believe 0.8 monkeys is too much for recycling via the console, would you consider halving that value?

**Comment by kriskog** [comment]
@Cyprex 
You want me to increase the manual recycling because it is too low and reduce the through console recycling because it is too high. I refuse to leave these in a state where they yield different numbers.

As for the returns, with t3 parts it would cost 0.4 monkeys per monkey after the change, as opposed to -0.5 if you are manually recycling on the current build. Don't even get me started on the comparison of t4.

So what are you actually after Cyprex? 0.3 on t1 - 0.6 on t4 parts is fine with me, but in that case I am changing it for both manual and console recycling.


Personally, I don't think the ratio I set is a problem. As I mentioned earlier the bottleneck of monkeys is before t3 parts currently, and still will be after the change. Reproductive and industrial greys are still very viable, since you need grey extracts and monkey cubes for crossbreeds regardless of how many cubes you got in the console.

**Comment by Cyprex** [comment]
>As for the returns, with t3 parts it would cost 0.4 monkeys per monkey after the change, as opposed to -0.5 if you are manually recycling on the current build. Don't even get me started on the comparison of t4.

Why compare manual recycling with console recycling? There are currently two ways to play: The labor-intensive way of dragging the monkeys into the recycler, and using the console. Obviously you'd want to compare current console usage(0.2) to future console usage (0.2-0.8).

>I refuse to leave these in a state where they yield different numbers.
My point in this was having the recycler good enough for the manual labor to be worth it, but it's your call.

**Comment by ExcessiveUseOfCobblestone** [comment]
There's so many options to compound with the recycler I don't see an issue with it getting nerfed.

Even with T4 parts on console you will eventually run out unless you use another method, which I think is ideal.

**Pull request merged**
